mixer/pkg/api/grpcServer.go

```diff
@@ -256,64 +256,67 @@ func (s *grpcServer) Report(ctx context.Context, req *mixerpb.ReportRequest) (*m
+	var accumBag *attribute.MutableBag
+	var reportBag *attribute.MutableBag
+
+	totalBags := len(req.Attributes)
+	for i := range req.Attributes {
+		lg.Debugf("Dispatching Report %d out of %d", i+1, totalBags)
 		span, newctx := opentracing.StartSpanFromContext(reportCtx, fmt.Sprintf("attribute bag %d", i))
 
-		// the first attribute block is handled by the protoBag as a foundation,
-		// deltas are applied to the child bag (i.e. requestBag)
-		if i > 0 {
-			if err := accumBag.UpdateBagFromProto(&req.Attributes[i], s.globalWordList); err != nil {
-				err = fmt.Errorf("request could not be processed due to invalid attributes: %v", err)
+		switch req.RepeatedAttributesSemantics {
+		case mixerpb.DELTA_ENCODING:
+			if i == 0 {
+				protoBag = attribute.NewProtoBag(&req.Attributes[i], s.globalDict, s.globalWordList)
+				accumBag = attribute.GetMutableBag(protoBag)
+				reportBag = attribute.GetMutableBag(accumBag)
+			} else {
+				if err := accumBag.UpdateBagFromProto(&req.Attributes[i], s.globalWordList); err != nil {
+					err = fmt.Errorf("request could not be processed due to invalid attributes: %v", err)
+					span.LogFields(otlog.String("error", err.Error()))
+					span.Finish()
+					errors = multierror.Append(errors, err)
+					break
+				}
+			}
+			if err := dispatchSingleReport(newctx, s.dispatcher, reporter, accumBag, reportBag); err != nil {
 				span.LogFields(otlog.String("error", err.Error()))
 				span.Finish()
 				errors = multierror.Append(errors, err)
-				break
+				continue
+			}
+		case mixerpb.INDEPENDENT_ENCODING:
+			protoBag = attribute.NewProtoBag(&req.Attributes[i], s.globalDict, s.globalWordList)
+			reportBag = attribute.GetMutableBag(protoBag)
+			if err := dispatchSingleReport(newctx, s.dispatcher, reporter, protoBag, reportBag); err != nil {
+				span.LogFields(otlog.String("error", err.Error()))
+				span.Finish()
+				errors = multierror.Append(errors, err)
+				continue
 			}
-		}
-
-		lg.Debug("Dispatching Preprocess")
-
-		if err := s.dispatcher.Preprocess(newctx, accumBag, reportBag); err != nil {
-			err = fmt.Errorf("preprocessing attributes failed: %v", err)
-			span.LogFields(otlog.String("error", err.Error()))
-			span.Finish()
-			errors = multierror.Append(errors, err)
-			continue
-		}
-
-		lg.Debug("Dispatching to main adapters after running preprocessors")
-		lg.Debuga("Attribute Bag: \n", reportBag)
-		lg.Debugf("Dispatching Report %d out of %d", i+1, len(req.Attributes))
-
-		if err := reporter.Report(reportBag); err != nil {
-			span.LogFields(otlog.String("error", err.Error()))
-			span.Finish()
-			errors = multierror.Append(errors, err)
-			continue
 		}
 
 		span.Finish()
 
-		// purge the effect of the Preprocess call so that the next time through everything is clean
-		reportBag.Reset()
+		switch req.RepeatedAttributesSemantics {
+		case mixerpb.DELTA_ENCODING:
+			reportBag.Reset()
+		case mixerpb.INDEPENDENT_ENCODING:
+			reportBag.Done()
+			protoBag.Done()
+			reportBag = nil
```

Comment:
nit: seems unnecessary. not clear why `reportBag` is cleared but not `protoBag`

Reply:
iirc, the bag code has a check to see if you call `Done()` twice (which it doesn't allow). So, I set it to `nil` in case it falls out to to the other block.

Which is why I added that check for nil in the code for `Done()`. 

I could not set it to nil, but then I'd have to put a guard in the code later.

Which is better for readability?

Reply:
I think we should not call Done twice. We can do that with a guard around outer Done.

Reply:
added.